examples/shader/gpu_readback.rs

```diff
@@ -0,0 +1,282 @@
+            push_constant_ranges: Vec::new(),
+            shader: shader.clone(),
+            shader_defs: Vec::new(),
+            entry_point: "main".into(),
+        });
+
+        ComputePipeline { layout, pipeline }
+    }
+}
+
+fn map_and_read_buffer(
+    render_device: Res<RenderDevice>,
+    buffers: Res<Buffers>,
+    sender: Res<RenderWorldSender>,
+) {
+    // Finally time to get our data.
+    // First we get a buffer slice which represents a chunk of the buffer (which we
+    // can't access yet).
+    // We want the whole thing so use unbounded range.
+    let buffer_slice = buffers.cpu_buffer.slice(..);
+    // Now things get complicated. WebGPU, for safety reasons, only allows either the GPU
+    // or CPU to access a buffer's contents at a time. We need to "map" the buffer which means
+    // flipping ownership of the buffer over to the CPU and making access legal. We do this
+    // with `BufferSlice::map_async`.
+    //
+    // The problem is that map_async is not an async function so we can't await it. What
+    // we need to do instead is pass in a closure that will be executed when the slice is
+    // either mapped or the mapping has failed.
+    //
+    // The problem with this is that we don't have a reliable way to wait in the main
+    // code for the buffer to be mapped and even worse, calling get_mapped_range or
+    // get_mapped_range_mut prematurely will cause a panic, not return an error.
+    //
+    // Using channels solves this as awaiting the receiving of a message from
+    // the passed closure will force the outside code to wait. It also doesn't hurt
+    // if the closure finishes before the outside code catches up as the message is
+    // buffered and receiving will just pick that up.
+    //
+    // It may also be worth noting that although on native, the usage of asynchronous
+    // channels is wholly unnecessary, for the sake of portability to WASM
+    // we'll use async channels that work on both native and WASM.
+
+    let (s, r) = crossbeam_channel::bounded::<()>(1);
+
+    // Maps the buffer so it can be read on the cpu
+    buffer_slice.map_async(MapMode::Read, move |r| match r {
+        // This will execute once the gpu is ready, so after the call to poll()
+        Ok(_) => s.send(()).expect("Failed to send map update"),
+        Err(err) => panic!("Failed to map buffer {err}"),
+    });
+
+    // In order for the mapping to be completed, one of three things must happen.
+    // One of those can be calling `Device::poll`. This isn't necessary on the web as devices
+    // are polled automatically but natively, we need to make sure this happens manually.
+    // `Maintain::Wait` will cause the thread to wait on native but not on WebGpu.
+
+    // This blocks until the gpu is done executing everything
+    render_device.poll(Maintain::wait()).panic_on_timeout();
+
+    // This blocks until the buffer is mapped
+    r.recv().expect("Failed to receive the map_async message");
```

Comment:
So this one is only necessary for WASM, given that the previous line blocks on native?

Reply:
Yes, it works on native without this channel, but it's safer to use it anyway. Wgpu only guarantees that buffers are mapped once the callback from map_async is sent. The fact that the map is done after the poll is not a guarantee.